### DISCUSSION TITLE
Align nuget package version and dll versions

### DISCUSSION THREAD

**Discussion by ReneLombard** [dotnet/runtime#109012] [General] [answered]
Is it possible to align the dll version and nuget packages on the same version, forexample, if you download the **System.Text.Json** nuget version: 8.0.5
![image](https://github.com/user-attachments/assets/0101bdff-8d53-4a31-bd6c-ef5ce6c82a5c)

You have the dll assembly version of 8.0.1024.46610
![image](https://github.com/user-attachments/assets/74b6db10-02e4-4608-852f-ff43552936cb)

**Comment by huoyaoyuan** [comment]
AssemblyFileVersion is NOT *assembly version*. They are different concepts.

Different version numbers follows different requirements. For example, some component need to be unique numeric value for each nightly build. Some allows strings while some allows numbers only.

A same looking string can't satisfy all the different requirements.

**Reply by ReneLombard** [reply]
Thank you for the explanation, 

I'm assuming the Nuget packages version follows [semantic versioning](https://semver.org/) where the Assembly version uses the [System versioning](https://learn.microsoft.com/en-us/dotnet/api/system.version?view=net-8.0) 

However, having multiple versions that produces the same DLLs isn't inherently obvious what the differences are, nightly build numbers could potentially be a postfix after the base version, to add more confusion to the mix in windows assembly versions are mapped to different fields,

![image](https://github.com/user-attachments/assets/9eaae124-6ba9-427a-92b8-ce7de3d7d9cb)

Assembly version vs. Assembly File Version vs. File Version vs. Product version

**Reply by huoyaoyuan** [reply, marked as answer]
The current situation is already the best effort. The version numbers are used by different subsystems with conflicting rules.

> to add more confusion to the mix in windows assembly versions are mapped to different fields,

> Assembly version vs. Assembly File Version vs. File Version vs. Product version

To get less confusing:

*Assembly version* is a specialized terminology. It's used by the CLR loader and has strong compatibility requirement.
File Version is `[AssemblyFileVersion]` and Product Version is `[AssemblyInformationalVersion]`. These two attributes are direct mapping of Win32 resource. The former is numeric and the latter is string. `[AssemblyInformationalVersion]` a.k.a. Product Version is directly tracible to individual commits. It's the only field that's not sortable.

**Reply by ReneLombard** [reply]
@huoyaoyuan Thank you for your clear explanation,
